ddtrace/appsec/processor.py

```diff
@@ -0,0 +1,119 @@
+import errno
+import json
+import os.path
+from typing import Any
+from typing import ClassVar
+from typing import Optional
+from typing import TYPE_CHECKING
+
+import attr
+
+import ddtrace
+from ddtrace import config
+from ddtrace.appsec._ddwaf import DDWaf
+from ddtrace.constants import MANUAL_KEEP_KEY
+from ddtrace.internal.logger import get_logger
+from ddtrace.utils.formats import get_env
+
+
+if TYPE_CHECKING:
+    from ddtrace import Span
+
+ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
+DEFAULT_RULES = os.path.join(ROOT_DIR, "rules.json")
+
+log = get_logger(__name__)
+
+
+@attr.s(eq=False)
+class AppSecProcessor(object):
+
+    tracer = attr.ib(type=ddtrace.Tracer, default=ddtrace.tracer)
+    rules = attr.ib(type=str, factory=lambda: get_env("appsec", "rules", default=DEFAULT_RULES))
+    _ddwaf = attr.ib(type=DDWaf, default=None)
+
+    enabled = False  # type: ClassVar[bool]
+    _instance = None  # type: ClassVar[Optional[AppSecProcessor]]
+
+    def __attrs_post_init__(self):
+        # type: () -> None
+        if self._ddwaf is None:
+            try:
+                with open(self.rules, "r") as f:
+                    rules = json.load(f)
+            except EnvironmentError as err:
+                # DDAS-0001-03
+                if err.errno == errno.ENOENT:
+                    log.error("AppSec could not read the rule file %s. Reason: file does not exist", self.rules)
+                else:
+                    # DDAS-0001-03 - TODO: try to log reasons
+                    log.error("AppSec could not read the rule file %s.", self.rules)
+                raise
+            except json.decoder.JSONDecodeError:
+                # DDAS-0001-03
+                log.error("AppSec could not read the rule file %s. Reason: invalid JSON file", self.rules)
+                raise
+            except Exception:
+                # DDAS-0001-03 - TODO: try to log reasons
+                log.error("AppSec could not read the rule file %s.", self.rules)
+                raise
+            try:
+                self._ddwaf = DDWaf(rules)
+            except ValueError:
+                # Partial of DDAS-0005-00
+                log.warning("WAF initialization failed")
+                raise
+
+    @classmethod
+    def enable(cls, *args, **kwargs):
+        # type: (Any, Any) -> None
+        if cls._instance is not None:
+            return
+
+        try:
+            processor = cls(*args, **kwargs)
+        except Exception:
+            log.warning("AppSec module failed to load.", exc_info=True)
+            if config._raise:
+                raise
+            return
+
+        # Automatically enable the AppSec processor on the tracer
+        processor.tracer.on_start_span(processor._attach_web_span)
+        cls._instance = processor
+        cls.enabled = True
```

Comment:
```python
@property 
def enabled(self):
    return cls._instance is not None
```

Reply:
This would have to be a "class" property, like what we do with `RuntimeMetrics` in `ddtrace/runtime/__init__.py`, I suppose

Reply:
I made the move to SpanProcessor as suggested. Result is [here](https://github.com/DataDog/dd-trace-py/pull/3175/files) this should remove this point